wal/wal.go

```diff
@@ -202,6 +209,55 @@ func tryRepairWAL(path string) error {
 	return nil
 }
 
+type walState uint32
+
+const (
+	walStateUnknown walState = iota
+	// walStateInit is the state of the WAL when it is initialized. This means
+	// the WAL directory has been read and the necessary in-memory state
+	// initialized.
+	walStateInit
+	// walStateReplaying is the state of the WAL when it is replaying entries in
+	// its WAL directory. It is not allowed to log new entries in this state or
+	// to transition from walStateWriting to walStateReplaying.
+	walStateReplaying
+	// walStateWriting is the state during normal usage of the WAL, i.e. when
+	// new entries are written to the log.
+	walStateWriting
+)
+
+var validStateTransitions = map[walState]map[walState]struct{}{
+	walStateUnknown: {
+		walStateInit: {},
+	},
+	walStateInit: {
+		walStateReplaying: {},
+		walStateWriting:   {},
+	},
+	walStateReplaying: {
+		walStateWriting: {},
+	},
+	walStateWriting: {},
+}
+
+func (w *FileWAL) stateTransition(newState walState) error {
+	var curState walState
+	for {
+		curState = walState(w.state.Load())
+		if curState == newState {
+			return nil
+		}
+		if ok := w.state.CompareAndSwap(uint32(curState), uint32(newState)); ok {
```

Comment:
does this need to be atomic? Could we instead have the wal move it's state to ready/writing once replay has finished? So that way it only has a single call

Reply:
We could, but we don't necessarily replay a WAL when starting up so we also need a mechanism to avoid replays if we have written at least once. I agree that this is not perfect but not sure what else we could do to support both cases.

Reply:
What if we changed that? Instead of `ReplayWALs` being a public function that users need to call that it's apart of creating a new columnstore. That way it can always move the WAL through the state machine

Reply:
I think I suggested this a couple of weeks ago (what you suggest would be my preferred approach) but there was a bit of push back from @brancz and @metalmatze due to wanting to be able to indicate that the grpc server is healthy before replaying, but not ready until then. Should we revisit?

Reply:
Removed this state commit and tracking the change to the API in #328 since it should probably be a separate PR.

Reply:
Just for completeness as this has been proposed on Discord:
If we can get the approach of registering the ColumnStore server at a later time to an existing gRPC server to work we should be fine with replaying during creating the column store.